File: sdk/java/openai-chat/src/main/java/com/gonkagate/examples/ErrorHandler.java
package com.gonkagate.examples;

final class ErrorHandler {
    private ErrorHandler() {
    }

    static void handleRequestError(Exception error) {
        Integer statusCode = findStatusCode(error);

        if (statusCode != null) {
            switch (statusCode) {
                case 401 -> fail("401 Unauthorized. Check your API key.");
                case 402 -> fail("402 Payment Required. Check your GonkaGate balance or billing status.");
                case 429 -> fail("429 Too Many Requests. Slow down request rate and retry.");
                case 503 -> fail("503 Service Unavailable. Retry in a few seconds.");
                default -> {
                    // Fall through to generic error message below.
                }
            }
        }

        String message = Strings.firstNonBlank(error.getMessage(), "Unknown request error.");
        fail(message);
    }

    static void fail(String message) {
        System.err.println("Error: " + message);
        System.exit(1);
    }

    private static Integer findStatusCode(Throwable error) {
        for (Throwable current = error; current != null; current = current.getCause()) {
            Integer direct = readStatusCode(current);
            if (direct != null) {
                return direct;
            }

            Object response = readProperty(current, "response");
            Integer responseCode = readStatusCode(response);
            if (responseCode != null) {
                return responseCode;
            }
        }

        return null;
    }

    private static Integer readStatusCode(Object source) {
        Integer statusCode = readIntProperty(source, "statusCode");
        if (statusCode != null) {
            return statusCode;
        }

        return readIntProperty(source, "status");
    }

    private static Integer readIntProperty(Object source, String propertyName) {
        if (source == null) {
            return null;
        }

        try {
            Object value = source.getClass().getMethod(propertyName).invoke(source);
            if (value instanceof Number number) {
                return number.intValue();
            }

            if (value != null) {
                Object intValue = value.getClass().getMethod("value").invoke(value);
                if (intValue instanceof Number number) {
                    return number.intValue();
                }
            }
        } catch (Exception ignored) {
            // Ignore missing methods and parse failures.
        }

        return null;
    }

    private static Object readProperty(Object source, String propertyName) {
        if (source == null) {
            return null;
        }

        try {
            return source.getClass().getMethod(propertyName).invoke(source);
        } catch (Exception ignored) {
            return null;
        }
    }
}

File: sdk/csharp/openai-streaming/ErrorHandler.cs
using System.Net;

namespace GonkaGate.Examples;

internal static class ErrorHandler
{
    public static int HandleRequestException(Exception ex)
    {
        var status = GetStatusCode(ex);

        if (status == 401)
        {
            return Fail("401 Unauthorized. Check your API key.");
        }

        if (status == 402)
        {
            return Fail("402 Payment Required. Check your GonkaGate balance or billing status.");
        }

        if (status == 429)
        {
            return Fail("429 Too Many Requests. Slow down request rate and retry.");
        }

        if (status == 503)
        {
            return Fail("503 Service Unavailable. Retry in a few seconds.");
        }

        var message = GetErrorMessage(ex);
        return Fail(message);
    }

    public static int Fail(string message)
    {
        Console.Error.WriteLine($"Error: {message}");
        return 1;
    }

    private static int? GetStatusCode(Exception ex)
    {
        for (Exception? current = ex; current is not null; current = current.InnerException)
        {
            if (TryReadStatusCode(current, out var status))
            {
                return status;
            }

            if (TryReadResponseStatusCode(current, out status))
            {
                return status;
            }
        }

        return null;
    }

    private static bool TryReadStatusCode(object source, out int statusCode)
    {
        foreach (var propertyName in new[] { "StatusCode", "Status" })
        {
            if (TryReadIntProperty(source, propertyName, out statusCode))
            {
                return true;
            }
        }

        statusCode = default;
        return false;
    }

    private static bool TryReadResponseStatusCode(object source, out int statusCode)
    {
        var responseProperty = source.GetType().GetProperty("Response");
        var response = responseProperty?.GetValue(source);

        if (response is null)
        {
            statusCode = default;
            return false;
        }

        return TryReadStatusCode(response, out statusCode);
    }

    private static bool TryReadIntProperty(object source, string propertyName, out int value)
    {
        var property = source.GetType().GetProperty(propertyName);
        var rawValue = property?.GetValue(source);

        if (rawValue is int intValue)
        {
            value = intValue;
            return true;
        }

        if (rawValue is HttpStatusCode httpStatusCode)
        {
            value = (int)httpStatusCode;
            return true;
        }

        value = default;
        return false;
    }

    private static string GetErrorMessage(Exception ex)
    {
        for (Exception? current = ex; current is not null; current = current.InnerException)
        {
            var property = current.GetType().GetProperty("Message");
            var rawValue = property?.GetValue(current) as string;
            if (!string.IsNullOrWhiteSpace(rawValue))
            {
                return rawValue.Trim();
            }
        }

        return "Unknown request error.";
    }
}
